### ISSUE TITLE
[vm/ffi] Support asynchronous callbacks

### ISSUE THREAD

**Issue by sjindel-google** [dart-lang/sdk#37022]
**Update Sept 6, 2023**: Asynchronous callbacks with void returns can now be done with  [`NativeCallable.listener`](https://api.dart.dev/stable/3.1.0/dart-ffi/NativeCallable-class.html) from Dart 3.1.

**Update March 12, 2020**: Asynchronous callbacks can be done through the native ports mechanism. See the samples in [samples/ffi/async](https://github.com/dart-lang/sdk/tree/master/samples/ffi/async) and corresponding [c code](https://github.com/dart-lang/sdk/blob/master/runtime/bin/ffi_test/ffi_test_functions_vmspecific.cc#L418-L877). In the future we would like to add a more concise syntax that requires less boilerplate code.

An asynchronous callback can be invoked outside the parent isolate and schedules a microtask there.

**Comment by dcharkes** [comment]
Asynchronous callbacks can be invoked on an arbitrary thread.

Synchronous callbacks cannot be invoked on an arbitrary thread because it would break the Dart language semantics: only one thread executing Dart per isolate.

Note that we need a solution for managing isolate lifetime when registering an asynchronous callback to be called from any thread in C. If the isolate died, and the callback is called, we have undefined behavior. A possible solution suggested by @mkustermann is to provide call-once callbacks, and keep the isolate alive until they are all called exactly once. Or we could have a counter or an arbitrary condition that we can check.

**Comment by dcharkes** [comment]
For inspiration, Kotlin requires the C library to be aware of Kotlin when doing non-main-thread-callbacks:

> If the callback doesn't run in the main thread, it is mandatory to init the Kotlin/Native runtime by calling kotlin.native.initRuntimeIfNeeded().

Source: https://github.com/JetBrains/kotlin-native/blob/master/INTEROP.md#callbacks

**Comment by sjindel-google** [comment]
It's not the thread which the callback is tied to, it's the isolate. So our callbacks may be invoked on any thread so long as that thread has called `Dart_EnterIsolate`.

**Comment by dcharkes** [comment]
> It's not the thread which the callback is tied to, it's the isolate. So our callbacks may be invoked on any thread so long as that thread has called `Dart_EnterIsolate`.

I presume `Dart_EnterIsolate` also (could) prevent isolates from being shutdown or killed.

**Comment by sjindel-google** [comment]
Another solution to managing lifetime is that the asynchronous callbacks need to be explicitly de-registered, and the isolate lives until all callbacks are de-registered.

Another challenge is knowing which `Isolate` the callback should run on. In AOT we have only one trampoline per target method, and there may be indefinitely many isolates from the same source at runtime.

**Comment by derolf** [comment]
Does it mean that calling back into Dart works as long as Dart_EnterIsolate is called before invoking the callback? How does the native code know that?

**Comment by dcharkes** [comment]
We have not made a design for this feature yet, we'll explore the possibilities when creating a design.

**Comment by derolf** [comment]
Okay, we have a workaround for it:
- Thread X wants to run callback
- Thread X adds the callback to a global list of pending callbacks
- Thread X send sigusr1 to the process
- Main Isolate watches sigusr1 (using ProcessSignal.watch) and gets woken up
- Main Isolate calls via FFI into a function that executes all pending callbacks

**Comment by dcharkes** [comment]
@derolf nice!

Would you like to provide a minimal sample and contribute it as a PR for others to learn from for the time being?

**Comment by derolf** [comment]
It didn’t work so far because Flutter itself is using SIGUSR1/2 for hot reloading.

However, I saw that a NativePort can be sent through FFI to the c-layer. Do you have any example how to send something to that nativePort from C?

**Comment by dcharkes** [comment]
@derolf, to my understanding the NativePort solution only works if you own the Dart embedder yourself.

cc @mkustermann

Linking https://github.com/flutter/flutter/issues/46887.

**Comment by derolf** [comment]
Okay, I implemented a way that works reliable and clean.

The queue:

On the native side, you need a threadsafe queue that allows to:
-- Enqueue callbacks from some thread that you want to execute in the main isolate (enqueue)
-- Have a function exposed through ffi that blocks until there are callbacks in the queue (wait)
-- Have a function exposes through ffi that executes pending callbacks (execute)

The dance:

Now, you spawn a slave isolate that waits on the queue and then seconds a message to the main isolate.

If some thread wants to dispatch a callback, you enqueue it instead of calling it. This will wakeup the slave isolate and send a message to main.

The main isolate receives this message and calls into ffi to execute all pending callbacks.

That's it. Less than 100 LOC to do it all.

**Comment by mkustermann** [comment]
> Have a function exposed through ffi that blocks until there are callbacks in the queue (wait)

That is slightly problematic in an event-loop based system. The blocking will prevent the isolate from processing any other messages.

We'll write an example on how this can be done already now with our `dart_native_api.h` and ports.

**Comment by derolf** [comment]
@mkustermann The problem is that none of the `dart_native_api.h` functions are available/exported in Flutter, so I can't use them. Happy to see your example how you make that work!

The "slave" isolate's one and only job is to do this waiting on the queue. It's doing nothing else. Here's its code:

```Dart
class _SlaveIsolateMessage {
  _SlaveIsolateMessage(this.port, this.isolate);
  final SendPort port;
  final int isolate;
}

static void _slaveIsolate(_SlaveIsolateMessage msg) {
    print("_slaveIsolate running");
    while (_dart_ffi_wait_for_callbacks(msg.isolate) != 0) {
      print("_slaveIsolate has callbacks");
      msg.port.send(1);
    }
    print("_slaveIsolate done");
    msg.port.send(0);
  }
```

The `isolate` int used to distinguish different native callback queues as we plan to have more than one "main" isolate.

**Comment by dcharkes** [comment]
The samples have landed, see referenced commits.

**Comment by insinfo** [comment]
Some status of when this will be available, I saw that the Realm team is just waiting for this feature to be able to launch a Realm for Flutter

**Comment by dcharkes** [comment]
Asynchronous callbacks can be today done through the native ports mechanism as illustrated in [samples/ffi/async](https://github.com/dart-lang/sdk/tree/master/samples/ffi/async).

In the future we would like to add a more concise syntax that requires less boilerplate code.

**Comment by katyo** [comment]
> Asynchronous callbacks can be today done through the native ports mechanism as illustrated in [samples/ffi/async](https://github.com/dart-lang/sdk/tree/master/samples/ffi/async).
> 
> In the future we would like to add a more concise syntax that requires less boilerplate code.

As I understand we still depends from APIs declared by _dart_native_api.h_ on the native side so this technique still does not usable with flutter.

**Comment by mkustermann** [comment]
> As I understand we still depends from APIs declared by dart_native_api.h on the native side
> so this technique still does not usable with flutter.

The `dart:ffi` library exposes the native api symbols now to Dart code via `NativeApi.postCObject` (as well as `NativeApi.closeNativePort`, `NativeApi.newNativePort`). Dart code can make a `ReceivePort`, obtain it's native port via `receivePort.sendPort.nativePort`, give that native port as well as the `NativeApi.postCObject` function pointer to C code. The C code can then post messages on the port, which Dart code can react on.

@katyo Would that work for you?

**Comment by jld3103** [comment]
@mkustermann Can you give a small example with the method you're mentioning? I need this for my app, because it's basically a deal breaker and I don't exactly understand how your method should work in code.
Thanks in advance

**Comment by dcharkes** [comment]
@jld3103 the samples are [here](https://github.com/dart-lang/sdk/tree/master/samples/ffi/async).

**Comment by katyo** [comment]
@mkustermann Thanks a lot for short explanation.
As I understand, the idea is exporting needed Dart API calls at runtime.

**Comment by yaminet1024** [comment]
`NativeApi.postCObject` cannot use in flutter stable by dart 2.7, has any way to replace this code?

**Comment by dcharkes** [comment]
> `NativeApi.postCObject` cannot use in flutter stable by dart 2.7, has any way to replace this code?
> @dcharkes

Hi @yaminet1024, you have to switch to the dev channel to use it. (Until Flutter 1.13 with Dart 2.8 is released.)

**Comment by katyo** [comment]
I sends native port id as _userdata_ to native function which calls _callback_ with this _userdata_. So I can access to it in my _callback_ from __Dart__ side.
I need send notification to this port back from Dart. How can I cast native port id back to the `SendPort` to do it (easy)?
Currently I see one only solution by calling `NativeApi.postCObject` from Dart but it looks ugly.

**Comment by derolf** [comment]
Just a quick note: The solution I proposed above works like a charm for us in a Flutter app.

**Comment by jld3103** [comment]
Can you link to your repo if it's opensource?

**Comment by derolf** [comment]
We have an implementation in a propitiatory repository, but I think the whole asynchronous callback machinery was integrated into https://github.com/heremaps/gluecodium

**Comment by mikeperri** [comment]
Hi all, I set up a simple example of a Flutter plugin with async methods that works via FFI instead of MethodChannels - https://github.com/mikeperri/flutter-native-callbacks-example

Not an expert in any of this stuff, so please let me know if there are any issues. Seems to work well though!

**Comment by ganeshrvel** [comment]
Anyone who is interested in using Dart FFI and Go together find it here: https://github.com/mraleph/go_dart_ffi_example

Thanks to @mraleph 🎉

**Comment by ganeshrvel** [comment]
@dcharkes How do we send a `struct` value from a native library to Dart via **nativeport**? Please add examples. 

I am trying to send response values, which is a `struct`, from my native library to Dart. The native library compiles successfully but the dart vm is crashing. Although if I try to send back a int64 value it works fine in the Dart side too.

I am using nativeport for the FFI communication and the native library is written in Go.

File url: https://github.com/ganeshrvel/squash_archiver/blob/feature/go-archiver/packages/archiver_ffi/native/archiver_lib/dart_api_dl/dart_api_dl.go#L54

Interestingly none of the below enum values were able to send the `struct` value back correctly while the `Dart_CObject_kInt64` was able to carry an `int64` value. 

```c
  Dart_CObject_kNull = 0,
  Dart_CObject_kBool,
  Dart_CObject_kInt32,
  Dart_CObject_kInt64,
  Dart_CObject_kDouble,
  Dart_CObject_kString,
  Dart_CObject_kArray,
  Dart_CObject_kTypedData,
  Dart_CObject_kExternalTypedData,
  Dart_CObject_kSendPort,
  Dart_CObject_kCapability,
  Dart_CObject_kUnsupported,
  Dart_CObject_kNumberOfTypes
```

**Comment by dcharkes** [comment]
> @dcharkes How do we send a `struct` value from a native library to Dart via **nativeport**? Please add examples.

You would send an `Int64` which holds the `address` of a `Pointer<MyStruct>`. You send the address from C/C++ by casting a `MyStruct*` to an `int64_t`, and you construct the `Pointer` with `fromAddress`, and then use `.ref` to access the struct.

Note that this way the memory is still managed by C, so you must communicate with C when to free the underlying memory.

In the future it might be possible to do move the ownership of the memory to Dart by using `TypedData`, but that requires the dart `MyStruct` class to be able to be backed by `TypedData` instead of `Pointer<MyStruct>` which is still [under development](https://dart-review.googlesource.com/c/sdk/+/140290).

**Comment by ganeshrvel** [comment]
@dcharkes Thanks for the reply. Actually I kind of figured out how to get it done.

If anyone else is looking for a way to connect Dart and Go via FFI (asynchronously using nativeport) and want to use structs then use this:

Go side
```go
package dart_api_dl

import "C"
import "unsafe"

// #include "stdlib.h"
// #include "stdint.h"
// #include "include/dart_api_dl.c"
//
// // Go does not allow calling C function pointers directly. So we are
// // forced to provide a trampoline.
// bool GoDart_PostCObject(Dart_Port_DL port, Dart_CObject* obj) {
//   return Dart_PostCObject_DL(port, obj);
// }
//
//	typedef struct Work{
//		int64_t a;
//		int64_t b;
//	}Work;
//
//	int64_t GetWork(void **ppWork) {
//		Work *pWork= (Work *)malloc(sizeof(Work));
//		pWork->a=16;
//		pWork->b=15;
//		*ppWork = pWork;
//
//		int64_t ptr = (int64_t)pWork;
//
//		return ptr;
//	}
import "C"

func Init(api unsafe.Pointer) C.long {
	return C.Dart_InitializeApiDL(api)
}

type Work struct {
	a int64
	b int64
}

func SendToPort(port int64, msg int64) {
	var obj C.Dart_CObject
	obj._type = C.Dart_CObject_kInt64

	var pwork unsafe.Pointer
	ptrAddr := C.GetWork(&pwork)

	*(*C.int64_t)(unsafe.Pointer(&obj.value[0])) = ptrAddr

	C.GoDart_PostCObject(C.int64_t(port), &obj)

	defer C.free(pwork)
}
```

Dart side:

```dart
    final interactiveCppRequests = ReceivePort();

    final interactiveCppSub = interactiveCppRequests.listen((data) {
      final work = Pointer<Work>.fromAddress(data as int);
      print(work.ref.a);
      print(work.ref.b);
    });

```

Find the full code here: https://github.com/ganeshrvel/squash_archiver/tree/feature/go-archiver-example/packages/archiver_ffi/native/archiver_lib


Thanks again.

**Comment by ganeshrvel** [comment]
@dcharkes Although I have a few questions about the **nativeport** approach, it's sort of the best practices:

1) Do we need to initialize ReceivePort and **nativeport** every time we make a FFI request or is it better to keep the native library instance in memory until the app is closed?
2) For two different FFI calls is it advisable to use a single **nativeport** and handle the messages from Go separately?

Cheers!

**Comment by dcharkes** [comment]
> Do we need to initialize ReceivePort and nativeport every time we make a FFI request or is it better to keep the native library instance in memory until the app is closed?

It is cleaner to keep it in memory, but requires a bit more book keeping.

> For two different FFI calls is it advisable to use a single nativeport and handle the messages from Go separately?

If the messages are going to the same isolate, it does not matter. Dart execution is single threaded, so the message handling is never concurrent. (Whatever makes your code the cleanest. I've not written lots of code that way, so I cannot advise on on which route to go.)

If the messages are going to different isolates you have to have two different native ports by definition.

**Comment by ganeshrvel** [comment]
>You would send an Int64 which holds the address of a Pointer<MyStruct>. You send the address from C/C++ by casting a MyStruct* to an int64_t, and you construct the Pointer with fromAddress in Dart, and then use .ref to access the struct.

Is there a way to send back a **dynamic** string array to Dart from the native library via nativeports? 

I was not able to find much resources on support for vectors in Dart FFI.

**Comment by dcharkes** [comment]
> > You would send an Int64 which holds the address of a Pointer. You send the address from C/C++ by casting a MyStruct* to an int64_t, and you construct the Pointer with fromAddress in Dart, and then use .ref to access the struct.
> 
> Is there a way to send back a **dynamic** string array to Dart from the native library via nativeports?
> 
> I was not able to find much resources on support for vectors in Dart FFI.

Vectors are a C++ feature (see https://github.com/dart-lang/sdk/issues/38788), or do you mean something else?

Please keep this issue on asynchronous callbacks, for other questions please open new issues (feel free to tag me).

**Comment by derolf** [comment]
I would strongly recommend to look into https://github.com/heremaps/gluecodium

It allows to create bindings for Dart, Java, Swift. I did a lot of work myself on the Dart bindings and we use it in a large scale project with thousands of functions.

**Comment by ganeshrvel** [comment]
> > Do we need to initialize ReceivePort and nativeport every time we make a FFI request or is it better to keep the native library instance in memory until the app is closed?
> 
> It is cleaner to keep it in memory, but requires a bit more book keeping.
> 
> > For two different FFI calls is it advisable to use a single nativeport and handle the messages from Go separately?
> 
> If the messages are going to the same isolate, it does not matter. Dart execution is single threaded, so the message handling is never concurrent. (Whatever makes your code the cleanest. I've not written lots of code that way, so I cannot advise on on which route to go.)
> 
> If the messages are going to different isolates you have to have two different native ports by definition.


- Should we manually close the native port after use?
- Is there a way to dispose or close a nativeport?
- What happens if the nativeport or the ffi instance is disposed off even before the native library returns a value? Does it cause any sort of memory leak or crash?
- Will a native asynchronous task terminate automatically if the nativeport or the ffi instance is disposed or closed?

@dcharkes

**Comment by dcharkes** [comment]
Please see the documentation at:
* https://api.dart.dev/stable/2.9.0/dart-isolate/SendPort-class.html
* https://api.dart.dev/stable/2.9.0/dart-isolate/ReceivePort-class.html
* https://github.com/dart-lang/sdk/blob/master/runtime/include/dart_native_api.h

> Is there a way to dispose or close a nativeport?

Yes, there are methods in Dart for closing them in Dart, and there is `Dart_CloseNativePort` for ports created with `Dart_NewNativePort` 

> Do we need to manually close the native port after use?

Yes, just like an open file.

> What happens if the nativeport or the ffi instance is disposed off even before the native library returns a value? Does it cause any sort of memory leak or crash?
See documentation on `Dart_PostCObject`.

```
 * If true is returned, the message was enqueued, and finalizers for external
 * typed data will eventually run, even if the receiving isolate shuts down
 * before processing the message. If false is returned, the message was not
 * enqueued and ownership of external typed data in the message remains with the
 * caller.
```

> Will a native asynchronous task terminate automatically if the nativeport or the ffi instance is disposed or closed?

Not if the native port is closed if the message already has been delivered - the Dart code is not cancelled mid execution.

What do you mean with "ffi instance is disposed or closed"? Do you mean the isolate the Dart code is running on? When an isolate is shut down, no more messages sent to it will be processed.

**Comment by HungerDeng** [comment]
> > @dcharkes How do we send a `struct` value from a native library to Dart via **nativeport**? Please add examples.
> 
> You would send an `Int64` which holds the `address` of a `Pointer<MyStruct>`. You send the address from C/C++ by casting a `MyStruct*` to an `int64_t`, and you construct the `Pointer` with `fromAddress` in Dart, and then use `.ref` to access the struct.
> 
> Note that this way the memory is still managed by C, so you must communicate with C when to free the underlying memory.
> 
> In the future it might be possible to do move the ownership of the memory to Dart by using `TypedData`, but that requires the Dart `MyStruct` class to be able to be backed by `TypedData` instead of `Pointer<MyStruct>` which is still [under development](https://dart-review.googlesource.com/c/sdk/+/140290).

I want to return a complex object (like Image object, or jni's jobject) to dart. How to define the struct int the dart side? The ffi just support the basic native type.

**Comment by HungerDeng** [comment]
> **Update March 12, 2020**: Asynchronous callbacks can be done through the native ports mechanism. See the samples in [samples/ffi/async](https://github.com/dart-lang/sdk/tree/master/samples/ffi/async). In the future we would like to add a more concise syntax that requires less boilerplate code.
> 
> An asynchronous callback can be invoked outside the parent isolate and schedules a microtask there.

Where can I find the example's c/c++ code?

**Comment by dcharkes** [comment]
@HungerDeng at the moment you can return `Pointer<MyStruct>` where `class MyStruct extends Struct`. See  [tests/ffi_2/function_structs_test.dart](https://github.com/dart-lang/sdk/blob/master/tests/ffi_2/function_structs_test.dart) for example.

We're working on implementing passing structs by value directly. This is tracked in https://github.com/dart-lang/sdk/issues/36730.

**Comment by HungerDeng** [comment]
> @HungerDeng at the moment you can return `Pointer<MyStruct>` where `class MyStruct extends Struct`. See [tests/ffi_2/function_structs_test.dart](https://github.com/dart-lang/sdk/blob/master/tests/ffi_2/function_structs_test.dart) for example.
> 
> We're working on implementing passing structs by value directly. This is tracked in #36730.

I want to get the jobject*(jni) by ffi to dart. And the dart code's struct hold the Pointer<JObject> variate. Then I want to pass the Pointer<JObject> to the flutter engine. Do you have any suggestions about how to  `recreate the  jobject* in the c++(flutter engine) from the dart Pointer<JObject>` ?
Thank you in advance.

**Comment by derolf** [comment]
Please describe your actual problem. What do you want to do with the jobject from within Dart?

**Comment by lumarama** [comment]
> Where can I find the example's c/c++ code?

Actually I have the same question. I see many references to the following example how to receive async callback in Dart code from C/C++ code:
https://github.com/dart-lang/sdk/blob/master/samples/ffi/async/sample_async_callback.dart

While this is only Dart part - where is corresponding C/C++ part?

**Comment by ganeshrvel** [comment]
@lumarama is this what you are looking for: https://github.com/dart-lang/sdk/blob/master/runtime/bin/ffi_test/ffi_test_functions_vmspecific.cc ?

**Comment by lumarama** [comment]
> @lumarama is this what you are looking for: https://github.com/dart-lang/sdk/blob/master/runtime/bin/ffi_test/ffi_test_functions_vmspecific.cc ?

Yes! Thanks! I think this is exactly what I was looking for. BTW, what is Opaque type used in the sample_async_callback.dart? 

class Work extends Opaque {}

**Comment by lumarama** [comment]
I'm wondering if I can do the following. I mean - I know that I can. This is a working code. But may be there is some catch why I should avoid doing this. So basically I create an array of INTs in Dart and pass it to my C-code - number of threads started in C-code and it returns to Dart. Then I check the array I passed to C-code periodically in Dart code to know when all thread are finished. Obviously C-threads will set corresponding array element to 1 when thread is finished. This way I can wait C-code processing while not occupying's my main UI thread.
  
  var threadStatusesPtr = xffi.allocate<ffi.Uint8>(count: _numThreads);
  var threadStatuses = threadStatusesPtr.asTypedList(_numThreads);
  for (int i=0; i<threadStatuses.length; i++) {
    threadStatuses[i] = 0;
  }

  _applyFiltersNative(_numThreads, threadStatusesPtr, ...);

  while (threadStatuses.any((status) => status == 0)) {
    await Future.delayed(Duration(milliseconds: 10));
  }

The reason I'm asking, even though this approach seemed to work, is because once I implemented this approach I started getting random "out of memory" issues - see https://stackoverflow.com/questions/65750426/strange-flutter-app-crash-empty-screen-with-out-of-memory-but-memory-is-fine

**Comment by katyo** [comment]
@lumarama If I undertand right you updates list items from running threads.
I assume some kind of locking mechanism should be applied to that solution because you have concurrent access to memory from different threads.

**Comment by lumarama** [comment]
> @lumarama If I understand right you updates list items from running threads.
> I assume some kind of locking mechanism should be applied to that solution because you have concurrent access to memory from different threads.

Yes, I do. While I was hoping (since implementing synchronization between Dart and C code is hard if possible at all) that you don't really need to synchronize in this case because Dart is always reading, while C-threads are always writing each in a separate list item only. So in worst case Dart can read list item earlier than C code writes to it - which isn't a problem - since I will try to read it again in the next loop cycle.

**Comment by lumarama** [comment]
So I'm almost 100% sure that the issue above is due to async call of my C-function that start parallel threads and returns to Dart. When I call it synchronously - it never happens. When I call it synchronously I do the following in the end of the C-function - I wait all threads to finish before I return to Dart:
   
```
  for (int i=0; i<numThreads; i++) {
      if (pthread_join(_threads[i], NULL) != 0) {
        printf("FFI: Error joining thread\n");
      }
    }
```

When I call it asynchronously I don't wait and return. Later (after a few minutes and many successful calls to this function) I start getting the following errors - as if there is memory leak somewhere:

pthread_create failed: couldn't mprotect R+W 1028096-byte thread mapping region: Out of memory

**Comment by vaind** [comment]
@lumarama Looks like you're not freeing the memory/closing threads for some reason. Maybe try running your code with valgrind:
```shell-script
dart2native ./test/some_test.dart -o /tmp/t
valgrind --show-mismatched-frees=no --leak-check=full /tmp/t
```

And maybe try moving this question to stackoverflow - you may have better luck with help there.

**Comment by lumarama** [comment]
I think I found what was the problem. By default pthread_create() creates joinable threads. That means you have to join them with pthread_join() - overwise they will never release resources allocated. That's why I've never seen this problem with synchronous call - because I waited all my threads with join. When you call your C-func asynchronously you don't want to wait your threads with pthead_join() - in this case you have to create detached threads - i.e. call pthread_detach() after pthread_create() - this way your threads won't hold any resources when finished.

**Comment by alexmercerind** [comment]
Are there any updates on this issue?

**Comment by derolf** [comment]
@alexmercerind 

> Are there any updates on this issue?

Have a look at https://github.com/heremaps/gluecodium . It supports Dart and also async callbacks.

**Comment by scvzerng** [comment]
My Windows dynamic library is a third party that cannot be modified. Is there a better way to support it？ thanks

**Comment by dcharkes** [comment]
> My Windows dynamic library is a third party that cannot be modified. Is there a better way to support it？ thanks

Not yet.

**Comment by scvzerng** [comment]
> > My Windows dynamic library is a third party that cannot be modified. Is there a better way to support it？ thanks
> 
> Not yet.
> 
> Note that you could load in another dll which you write yourself and has the required callback signature in C. And then you can use the Native Ports in that dll to call back asynchronously to Dart. And then you pass the function from your own dll as a callback to the library you cannot modify.

I know I could do that, but it just feels a little weird. Haha, thank you very much for your reply

**Comment by tatumizer** [comment]
Can someone explain to me what will happen if I simply pass dart's function pointer to C, and then C program will call it at any time?
Then, in the receiving end, the corresponding dart function will wait (how-TBD) until the my isolate becomes idle, execute the function and return the result to C. What's wrong with that?
The only thing I need is to be able to somehow implement "waiting for an isolate to become idle" part. Is it difficult? 
Impossible? Stupid? Am I missing something that makes the idea absurd?

**Comment by dcharkes** [comment]
It needs to not only be "waiting for an isolate to become idle", it also needs to be Dart executing thread, otherwise the isolate might wake up and start executing code (for example when it receives a message from a native port).

> Can someone explain to me what will happen if I simply pass dart's function pointer to C, and then C program will call it at any time?

It will crash your program (possibly in wild and wondrous ways).

> Impossible?

You can achieve it with writing an extra C function.

To achieve what you want:

1. Make a C function. This is the function you're going to pass to your other C functions as a function pointer.
2. That C function sends a message to Dart through the native ports. It leaves the actual thing that it wants to execute in some kind of queue, and then blocks execution through some concurrency mechanism. (Thread 1)
3. Dart receives the message, and calls from Dart into C, and executes the things in the queue doing the callbacks from C back in to Dart and leaves the results of the operations in the queue. (Thread 2: The main Dart execution thread.) And then signals Thread 1 that the results of the work is done, and then returns to Dart.
4. Thread 1 is now unblocked and can continue its C function returning the answer from the callback that Thread 2 left.

Here is the sample code: 

* https://github.com/dart-lang/sdk/blob/master/runtime/bin/ffi_test/ffi_test_functions_vmspecific.cc#L370-L828
* https://github.com/dart-lang/sdk/blob/master/samples/ffi/async/sample_async_callback.dart

**Comment by mraleph** [comment]
@tatumizer you forget that there can be multiple isolates in the system, so you need to know which one to use. When you call back synchronously we simply use the same isolate that is currently attached to the thread. But when you call back async it's unclear which isolate to use.

**Comment by tatumizer** [comment]
Oh I see. But what about the symmetric case? When I call C method from dart while C running in multi-threaded mode, will my call end up being executed in a random thread (whatever that means)? True, the threads in C are not isolates, they all share the same code, but ... no, I can't wrap my head around it. Maybe there's a standard pattern for that? Some Rendezvous mechanism or something? Or the solution @dcharkes outlined should be implemented on both sides?

**Comment by vaind** [comment]
For the lack of a better place to ask this question about a native DL API:

when sending a string message from C/C++ to Dart, it isn't clear if the message should be copied (malloc+memcpy) before sending or passed directly. 

```c++
void dartPostStringMessage(int64_t nativePort, const char* arg) {
    Dart_CObject obj;
    obj.type = Dart_CObject_kString;
    obj.value.as_string = const_cast<char*>(arg);
    Dart_PostCObject_DL(nativePort, &obj);
}
```

Originally, the type of the `as_string` member (`char*` instead of `const char*`) lead me to believe I should pass a copy of the string, but `valgrind` reported a leak when I expected dart to take over ownership of a `malloc`ed memory...

So... am I correct to assume that `Dart_PostCObject_DL()` expects an existing `const char*`. And also that it only accesses the string before returning to the caller (in other words, `arg` could be freed after `Dart_PostCObject_DL(nativePort, &obj);`)?

**Comment by dcharkes** [comment]
@vaind, feel free to open new issues for questions or use stack overflow.

The string is (decoded and) copied when send as message. So yes, it can be freed after `Dart_PostCObject` returns.

`Dart_PostCObject` implementation:

https://github.com/dart-lang/sdk/blob/82ea8416ed2a22e40c5c67252cb37bebec3a57e8/runtime/vm/native_api_impl.cc#L45-L60

`WriteCMessage ` implementation:

https://github.com/dart-lang/sdk/blob/82ea8416ed2a22e40c5c67252cb37bebec3a57e8/runtime/vm/dart_api_message.cc#L1018-L1061

**Comment by ganeshrvel** [comment]
@dcharkes Does the current version of Dart FFI support returning of a Struct from C?

Previously I got it working by returning the pointer of the Struct from C and later processing it inside the Dart. 

```C
	bool GoDart_PostCObject(Dart_Port_DL port, int64_t ptrAddr) {
  		Dart_CObject dartObj;
  		dartObj.type = Dart_CObject_kInt64;

		dartObj.value.as_int64 = ptrAddr;

	    return Dart_PostCObject_DL(port, &dartObj);
	}
```


A lot has changed in few months, I will have to rewrite a lot of the FFI related code now. 

- Is it possible to return a struct by value from C to dart using `nativeport`?
- If yes, is there any example or documentation that I can refer to?

**Comment by dcharkes** [comment]
> Is it possible to return a struct by value from C to dart using nativeport?

No it is not currently.

We could explore returning a TypedData through a native port, and exposing a constructor for structs that allows you to get it backed by a TypedData. (Currently you can only construct structs by FFI trampolines by value or from `Pointer`.) I've filed https://github.com/dart-lang/sdk/issues/46099 to track this.

**Comment by ganeshrvel** [comment]
> > Is it possible to return a struct by value from C to dart using nativeport?
> 
> No it is not currently.
> 
> We could explore returning a TypedData through a native port, and exposing a constructor for structs that allows you to get it backed by a TypedData. (Currently you can only construct structs by FFI trampolines by value or from `Pointer`.) I've filed #46099 to track this.

Thank you. That's great!

I have a request though, a long standing one and I have made this multiple times as well.

Dart FFI is such a great and powerful feature, but what it lacks right now is a good documentation. The examples and implementations are scattered all over the GitHub repository example codes and issues. The existing Dart FFI documentation site just points the user to the GitHub example page.


It's very difficult to know about the available features, changes made to the FFI and migration to the newer version. It's difficult to keep track of each feature.

Would it be too early to put out a documentation for Dart FFI?

**Comment by mschwartz** [comment]
Seems to me you could send JSON through the native port and deserialize it in Dart.  It's not exactly sending a Struct as a result, but might be useful anyway.

**Comment by mschwartz** [comment]
I'm going from memory here, but I do remember that you can enter an Isolate from different threads, just that only one thread at a time can enter any one Isolate.  An Isolate is not entered once you're in C/C++ code, right?

Does that help out?

**Comment by ro99** [comment]
@dcharkes I noticed that the [Dart_PostInteger](https://github.com/dart-lang/sdk/blob/f4180854c71cbec340f66fd1c605dc58f70ea38e/runtime/include/dart_native_api.h#L127) from dart_native_api.h is not exposed for the [NativeApi](https://github.com/dart-lang/sdk/blob/f4180854c71cbec340f66fd1c605dc58f70ea38e/sdk/lib/ffi/ffi.dart#L762) class. Would it be possible to expose it as well ?

**Comment by dcharkes** [comment]
@ro99, we've exposed a whole list of `dart_api.h` functions through `dart_api_dl.h`, including [`Dart_PostInteger`]. Use [`initializeApiDLData`] from `NativeApi` instead to initialize all the symbols from `dart_api_dl.h`.

[`Dart_PostInteger`]: https://github.com/dart-lang/sdk/blob/f4180854c71cbec340f66fd1c605dc58f70ea38e/runtime/include/dart_api_dl.h#L54
[`initializeApiDLData`]: https://github.com/dart-lang/sdk/blob/f4180854c71cbec340f66fd1c605dc58f70ea38e/sdk/lib/ffi/ffi.dart#L802

**Comment by derolf** [comment]
> without async function call or callback ,dart ffi could be less valuable

Make a concrete C-API proposal and the dart team will for sure consider it.

**Comment by dvc94ch** [comment]
@Sunbreak: is the libwrapper.so something that could be abstracted into a reusable component? So you register your dart closure with libwrapper which returns a function pointer which is passed to the native library?

**Comment by Keithcat1** [comment]
Why is it a problem that an asynchronous Dart FFI callback could mutate global state? Any C FFI is either unsafe or rather limited. Go allows C to invoke callbacks on another thread, even though Go has a multithreaded garbage collector. I have a feeling that trying to wrap low latency audio libraries like Miniaudio with a C callback that uses a send port to let the main Dart isolate would add some latency. Having to ship extra dynamic libraries in order to wrap FFI callbacks along with my application is also something I'd like to avoid if at all possible.

**Comment by blaugold** [comment]
Here’s a proposal for an API for async callbacks, that I think is generally useful.

```dart
/// A callback, that allows native code to call a static Dart function
/// asynchronously.
///
/// Any callback, for which [close] has not been called, will keep the
/// [Isolate] that created it alive.
abstract class AsyncCallback<T extends Function> {
  /// Creates a callback, that allows native code to call a Dart function
  /// asynchronously and blocks the native caller until the Dart function
  /// returns.
  ///
  /// If an exception is thrown while calling [f], the native function will
  /// return [exceptionalReturn], which must be assignable to return type of
  /// [f].
  AsyncCallback(
    @DartRepresentationOf("T") Function f, [
    Object? exceptionalReturn,
  ]);

  /// Creates a callback, that allows native code to call a Dart function
  /// asynchronously and in a non-blocking way.
  ///
  /// The native caller needs to ensure, that the arguments it passes to
  /// [nativeFunction] stay valid until the Dart function returns, and must
  /// not rely on the side effects of the Dart function.
  ///
  /// The [nativeFunction] always returns [nonBlockingReturn], which must be
  /// assignable to the return type of [f]. The return value of the Dart
  /// function is ignored.
  AsyncCallback.nonBlocking(
    @DartRepresentationOf("T") Function f, [
    Object? nonBlockingReturn,
  ]);

  /// C function pointer to a C function, that will call the Dart function.
  ///
  /// The C function can be called from any thread.
  ///
  /// Calling the C function is undefined behavior, after [close] has been
  /// called or the [Isolate] that created the callback has died.
  Pointer<NativeFunction<T>> get nativeFunction;

  /// Closes this callback.
  ///
  /// After this method has been called, [nativeFunction] must not be called
  /// anymore.
  ///
  /// A callback, that has not been closed, will keep the [Isolate] that created
  /// it alive.
  void close();
}
```

Non-blocking `AsyncCallback`s can be useful, if the native caller just notifies of some event or delivers a message and blocking it is detrimental to performance.

It would be useful for non-blocking `AsyncCallback`s to allow `CObject`s in the arguments of the callback.

With this proposal, it would be the responsibility of the Dart code, that created an `AsyncCallback`, to close it and ensure native code stops calling it, before that point.

It does not specify what happens when multiple `AsyncCallbacks` are created with the same static Dart function.

**Comment by dcharkes** [comment]
I was thinking of just adding named arguments `bool async`, `bool blocking` to `fromFunction` rather than introducing a new API.

Can you elaborate on your reasoning for a separate API?

> It would be useful for non-blocking AsyncCallbacks to allow CObjects in the arguments of the callback.

Why use `CObject`s instead of `Dart_Handle` in native and `Object` in Dart? That is how we do it for all the other FFI calls and callbacks. (The fact that it it could use native ports as an implementation mechanism should not leak out.)

> With this proposal, it would be the responsibility of the Dart code, that created an AsyncCallback, to close it and ensure native code stops calling it, before that point.

I would give async callbacks the same semantics as synchronous callbacks for now: once allocated they stay alive until the isolate terminates. We should indeed also add an API at some point that can reclaim the memory used for the callbacks. It would indeed be the responsibility of the native code not to call those function pointers anymore after that.

**Comment by blaugold** [comment]
> Can you elaborate on your reasoning for a separate API?

I was thinking of an `AsyncCallback` as something similar to a `ReceivePort`, which prevents the isolate from exiting while it is not closed. If an isolate does nothing else but create an async callback, pass it to native code and wait for the callback to be called, what would stop the isolate from exiting? The isolate could schedule a timer far in the future and cancel it when it's done receiving callbacks, but that seems a bit like a hack.

> Why use CObjects instead of Dart_Handle in native and Object in Dart? That is how we do it for all the other FFI calls and callbacks. (The fact that it it could use native ports as an implementation mechanism should not leak out.)

I was thinking about how a non-blocking caller would be able to pass data that needs to be allocated to an async callback.

Let's take a log callback as an example. It doesn't have to be blocking because the caller just wants to hand a message and maybe a few other values over to Dart code. When native code passes a `Dart_CObject` to `Dart_PostCObject`, it is deeply copied and does not need to remain allocated, even though Dart code receives that value asynchronously. 

The native Dart API that is currently available through `dart_api_dl.h`, doesn't allow native code to build a `Dart_Handle` to a `String` or `List`, for example.

---

BTW, looking forward to acdf82de1714a6a16c5f604e233c990588b98a4c being available in one of the release channels.
Looks great.

**Comment by dcharkes** [comment]
> which prevents the isolate from exiting while it is not closed

It would by default not prevents isolates from exiting. Because without a cancel API the isolates would never exit. Also, it's easy to miss.

Instead, I would suggest opening another port from your code, and sending an exit message to that on the last callback from native code, to prevent the isolate from exiting until native code is done with it.

You could implement `AsyncCallback` as a user like this. (Well, except the fact that we're not supporting generics in FFI callback signatures, so not fully.)

The lack of support for String (https://github.com/dart-lang/sdk/issues/39787) and Lists is a separate issue in my opinion. We would also like to support these for FFI calls and synchronous FFI callbacks, they are not async callback specific. (Side note: For creating a list in native code, we can do callbacks with handles to build the list. Both the `dart_api` and FFI callbacks do transitions from native code in the VM, so the performance should be better with FFI as it is not reflective. Side note 2: Same for creating a Dart string, we can do a callback that passes in a Pointer, and that gets the string as a return value in a handle.)

**Comment by blaugold** [comment]
Thanks for the explanation. Makes sense to keep the API simple and consistent.

**Comment by mraleph** [comment]
I have been thinking that https://github.com/dart-lang/sdk/issues/47778 might also provide a way to tackle this issue to a degree. Isolate independent functions can be invoked on any thread and then could communicate back to the (isolate dependent) Dart world via ports. I will incorporate these considerations on the initial design.

**Comment by dvc94ch** [comment]
Cool. Does this mean the flutter engine can be just a regular dynamic library once this issue gets resolved?

**Comment by maks** [comment]
@mraleph Having any kind of better support for callbacks in FFI would be a big win. I find that dealing with callbacks from C  by **far** the most difficult thing and roadblock to using FFI with many C libraries. I completely understand that there is the restriction due to the 1 thread per Isolate at a time rule, but that doesn't make life any easier when using FFI. 

Looking at the Finalizer proposal doc, it looks great for non-Isolate code, but could it for instance be expanded to something that would work with Isolates? I was thinking maybe it could be expanded by having functionality that creates a new Isolate per each native callback into Dart, thus ensuring that it meets the re-entrancy restriction? With the recent improvements with Isoalate Groups when it comes to Isolate creation performance and mem overhead, could this be a way forward? I guess the main difficulty with "create a new Isolate per callback" approach would be how to wire up ports to existing Isolates to make these automatically created Isolates actually useful, but perhaps thats a more tractable problem and I would think is better than that current workaround of needing us Dart developers to deal with ports/Isolates in the native side code.

**Comment by mraleph** [comment]
@maks If the APIs you are finding challenging to bound to are public please share some details of what you are trying to achieve and where it is failing.

> I was thinking maybe it could be expanded by having functionality that creates a new Isolate per each native callback into Dart, thus ensuring that it meets the re-entrancy restriction? 

I have thought about this approach before but it raises a question: what are you going to do in that newly created isolate? It is not going to share any memory with another isolate so callback essentially has to be stateless. And if it is stateless then isolate independent code is probably a good enough answer. Alternatively if you are planning to use that freshly created isolate just as a jumping point to another isolate - by sending some data through `SendPort` then having automatic way to marshal asynchronous invocations to another isolate is good enough answer as well.

**Comment by Keithcat1** [comment]
The FFI finalizers proposal mentioned possibly allowing Dart code to be compiled into native callbacks so long as it only invokes C functions. This would probably allow sending messages to the main isolate to run the callback there without having to compile a separate DLL alongside your app.

**Comment by maks** [comment]
@mraleph good points! What I had in mind was really just moving the current workarounds use of `SendPort`s into the Dart side vs having to do it in native code. I don't quite follow what you mean by:
> having automatic way to marshal asynchronous invocations to another isolate

as reading the finalizers proposal, it talked about running outside any Isolate and only allowing Dart code that could only call FFI code and not reference any other Dart code. 

What I had in mind was some sort of mechanism to allow newly created Isolates to be able to get hold of a `SendPort` from another existing Isolate. Some sort of registry perhaps, so that other Isolates could register `SendPorts` there and then those SendPorts would be passed in via the `spawn()` that got called to create these "automatic" Isolates for native callbacks. 
For example
```dart
// this could be the main Isolate
final receivePort = ReceivePort();
portRegistry.add("sendMeNativeStuff",  receivePort.sendPort);
```

and then when the "automatic" Isolate is created:
```dart
Isolate.spawn(SendPort registeredPort) {
  registeredPort.send(thisIsSomeDataFromNative)
}
```

I'm not sure if I've explained, so hope the above makes sense?

In regards to which libraries I've run into problems with: its basically any library that has a async callback api. The first one I ran into a while back [was libfuse](https://libfuse.github.io/doxygen/example_2hello_8c.html). Since then I've also found that most audio playback libraries use callbacks as well, eg. [miniaudio](https://github.com/mackron/miniaudio) and my [initial attempt at a binding](https://github.com/maks/dainty_audio).

**Comment by mraleph** [comment]
> I don't quite follow what you mean by:
> 
> > having automatic way to marshal asynchronous invocations to another isolate

I mean we could provide a way to write code like this: 

```dart
// Create callback from a function [func] which supports being called 
// from a thread which does not have a Dart isolate associated with it.
// If such situation occurs it will instead send a message back to this
// isolate and wait for a response.
final cb = Pointer.fromFunction<Int32 Function(Int32)>(func, detached: true, synchronous: true);
```

It's not going to work well though for situations which require fast callbacks.

In this cases isolate independent code is likely a better answer. @dcharkes is working on some initial prototypes right now.

**Comment by roc2539** [comment]
> My Windows dynamic library is a third party that cannot be modified. Is there a better way to support it？ thanks
Is there a better way to support it？

**Comment by kenfred** [comment]
@dcharkes I have followed the code in samples/ffi/async. Calls to `Dart_PostCObject_DL` return true, but I never get an event on the dart side that is listening to the ReceivePort.

Do you have any tips on how to debug this?

**Comment by insinfo** [comment]
I was trying to make a Binding of the PortAudio library with dart FFI and I came across this problem, will it be possible in the future to make a Binding with dart FFI of C libraries as well as in Go without having to modify or create a Wrapping in c++ to get around the problem "Cannot invoke native callback outside an isolate." This is very frustrating.

Is there any prospect for the future for this issue to be resolved on the dart side so that I don't have to create another shared library?

```dart
import 'dart:convert';
import 'dart:ffi' as ffi;
import 'package:audio_player/portaudio.dart';
import 'package:ffi/ffi.dart' as ffi;
import 'package:path/path.dart' as path;
import 'dart:io' show Directory, exit;

late Portaudio portaudio;

const NUM_SECONDS = 5;
const double SAMPLE_RATE = 44100;
const FRAMES_PER_BUFFER = 64;
const M_PI = 3.14159265;
const TABLE_SIZE = 200;
var err = 0;
void main(List<String> args) {
  var libraryPath =
      path.join(Directory.current.path, 'bin', 'libportaudio64bit.dll');
  var dynamicLibrary = ffi.DynamicLibrary.open(libraryPath);
  portaudio = Portaudio(dynamicLibrary);
  print(portaudio.Pa_GetVersion());
  //init lib
  var result = portaudio.Pa_Initialize();
  checkForError('initialize', result);
  // configure output
  var size = ffi.sizeOf<PaStreamParameters>();
  final ffi.Pointer<PaStreamParameters> outputParameters =
      ffi.malloc.allocate(size);

  outputParameters.ref.channelCount = 2;
  outputParameters.ref.sampleFormat = SampleFormat.float32;
  outputParameters.ref.device = portaudio.Pa_GetDefaultOutputDevice();
  if (outputParameters.ref.device == -1) {
    print("Error: No default output device.\n");
  }
  outputParameters.ref.suggestedLatency =
      portaudio.Pa_GetDeviceInfo(outputParameters.ref.device)
          .ref
          .defaultLowOutputLatency;
  outputParameters.ref.hostApiSpecificStreamInfo = ffi.nullptr;

  // init stream
  final stream = 
 ffi. Pointer<ffi.Pointer<ffi.Void>>.fromAddress(ffi.malloc.allocate(ffi.sizeOf<ffi.Int32>()).address);
//final stream =ffi. Pointer<ffi.Pointer<ffi.Void>>.fromAddress(ffi.malloc.allocate(ffi.sizeOf<ffi.Int32>()).address);


  err = portaudio.Pa_OpenStream(
      stream,
      /* no input */
      ffi.nullptr,
      outputParameters,
      SAMPLE_RATE,
      FRAMES_PER_BUFFER,
      /* we won't output out of range samples so don't bother clipping them */
      paClipOff,
      ffi.Pointer.fromFunction<example_callback>(patestCallback, 1),
      ffi.nullptr);
  print(err);
  checkForError('Pa_OpenStream', err);

  err = portaudio.Pa_StartStream(stream.value);
  checkForError('Pa_StartStream', err);

  print("Play for $NUM_SECONDS seconds. ", );
  portaudio.Pa_Sleep(NUM_SECONDS * 1000);
}

int patestCallback(
    ffi.Pointer<ffi.Void> input,
    ffi.Pointer<ffi.Void> output,
    int frameCount,
    ffi.Pointer<PaStreamCallbackTimeInfo> timeInfo,
    int statusFlags,
    ffi.Pointer<ffi.Void> userData) {
  print('patestCallback');
  return PaStreamCallbackResult.paComplete;
}

void checkForError(String text, int result) {
  if (result != PaErrorCode.paNoError) {
    var errText = portaudio.getErrorText(result);
    print('Error $text - $errText');
    exit(1);
  }
}

class PaTestData extends ffi.Struct {
  @ffi.Array(TABLE_SIZE)
  external ffi.Array<ffi.Float> sine;

  @ffi.Int()
  external int left_phase;

  @ffi.Int()
  external int right_phase;

  @ffi.Array(20)
  external ffi.Array<ffi.Int8> message;
}

extension CharArray on ffi.Array<ffi.Int8> {
  String getDartString(int maxLength) {
    var list = <int>[];
    for (var i = 0; i < maxLength; i++) {
      if (this[i] != 0) list.add(this[i]);
    }
    return utf8.decode(list);
  }

  void setDartString(String s, int maxLength) {
    var list = utf8.encode(s);
    for (var i = 0; i < maxLength; i++) {
      this[i] = i < list.length ? list[i] : 0;
    }
  }
}


```

https://github.com/PortAudio/portaudio/blob/master/examples/paex_sine.c


```console
PS C:\MyProjectsDart\audio_player> dart .\bin\teste.dart
1246976
0
Play for 5 seconds. 
../../runtime/vm/runtime_entry.cc: 3766: error: Cannot invoke native callback outside an isolate.
version=2.18.3 (stable) (Mon Oct 17 13:23:20 2022 +0000) on "windows_x64"
pid=5224, thread=6900, isolate_group=(nil)(0000000000000000), isolate=(nil)(0000000000000000)    
isolate_instructions=0, vm_instructions=7ff7430566e0
  pc 0x00007ff743260132 fp 0x000000f28e9ff2f0 Dart_IsPrecompiledRuntime+0x21ee92
-- End of DumpStackTrace
PS C:\MyProjectsDart\audio_player> 
```

**Comment by xiaoxiaowesley** [comment]
> @dcharkes I have followed the code in samples/ffi/async. Calls to `Dart_PostCObject_DL` return true, but I never get an event on the dart side that is listening to the ReceivePort.
> 
> Do you have any tips on how to debug this?
> 
> **Edit** I found my issue. Here are some details in case it may help someone in the future..
> 
> Here is the basic mechanism from the sample code referenced above:
> 
> 
> In my case, the event was triggered from Dart and it blocked waiting for the callback to execute. This caused a deadlock and explains why the ReceivePort listen never seemed to receive the work.
> 
> 
> So the obvious fix is to break the interlock between Dart and Native when the event originates from Dart:

Any update? I met the same problem.

**Comment by kenfred** [comment]
> Any update? I met the same problem.

You can see in my post what was wrong and how I signed it.

**Comment by mkustermann** [comment]
> You can see in my post what was wrong and how I signed it.

We are actively looking into how to solve this atm. Stay tuned for more updates soon.

**Comment by liamappelbe** [comment]
I'm nearing completion of this feature. I expect to land it in the dev channel in around a month.

The way it will work is that there will be another method similar to `Pointer.fromFunction`, and when native code calls the function pointer it returns (from any thread), the arguments will be bundled into an inter-isolate message, and sent over an ordinary port to a handler that invokes the target Dart function in the same isolate that registered the callback. It's a fire and forget mechanism, so the native code doesn't wait for a response message. This means that only functions that return void will be supported.

**Comment by maks** [comment]
@liamappelbe likewise I'm also very interested in this support for the same use case as @insinfo: using FFI with low-latency audio libs that pretty much all work via callbacks. I noticed that the above commit seems to have added support for these so I'm guessing this is coming soon to Dart master channel?

**Comment by liamappelbe** [comment]
@maks Yeah, async callbacks have landed, so should be available on the master channel soon. If everything goes according to plan, they'll be in Dart 3.1. They're created using [`NativeCallable.listener`](https://github.com/dart-lang/sdk/blob/main/sdk/lib/ffi/ffi.dart#L182).

They're very async tho, so you might have issues using them for low latency audio. When the native code calls an async callback, all the args are bundled into a message that's sent to the target isolate to invoke the callback. The native code doesn't wait for the callback to be invoked, or wait for a response. It just sends the message then continues execution.

For example, one common pattern in audio APIs is that the API will invoke your callback when it's time to fill an audio buffer. The API might assume that by the time the callback returns, the buffer is filled. That assumption would not be valid if you used async callbacks for this.

Of course, it depends on the details of the API you're interacting with. So just keep the limitations in mind.

**Comment by littleGnAl** [comment]
@liamappelbe 
Is any plan to support the "native code waits for the callback to be invoked"? If it's async, I think a data copy may be needed, which is not a performance way to handle the low latency audio/video callback.

I think if the "native code waits for the callback to be invoked" is supported, it can unlock a new way to process the real-time audio/video callback on the dart side.

**Comment by maks** [comment]
@liamappelbe ah yes of course, you're quite right, miniaudio (the lib I had in mind) does indeed except the buffer to be filled on return so yes, this unfortunately won't help for that case.

**Comment by liamappelbe** [comment]
We considered a lot of different async callback designs before settling on `NativeCallable.listener`. Every approach had trade-offs, and there were no perfect solutions. We can add some of the other approaches if people need them.

@littleGnAl We want to add support for blocking callbacks like that, by having the caller thread enter the target isolate. The main problem is that the target isolate will typically already be entered by another thread, so the caller thread has to block and wait to enter the isolate (eg at a message loop boundary). This approach is a lot more risky, and it's possible for developers to put themselves in a deadlock situation. Also, I still don't think this would count as "low latency" enough for real time audio, since you won't know how long the caller thread will be blocked for while it waits to enter the target isolate. When we did a use case study, we found we could support a lot of use cases with the non-blocking `NativeCallable.listener` approach, which is much safer, so we decided to land that first.

@maks If you're trying to wrap miniaudio to present it as a Dart package, and allow users to write Dart code that fills audio buffers, you might need one of the other async callback approaches we considered: temporary isolates. In this model, async callbacks would be run on the caller thread, in a temporary isolate. This has all sorts of limitations, because the callback is running in a different isolate, so can't access/mutate the rest of the app state. But it's the only approach that can run Dart code on a random thread that will be really low latency. I suppose if there was a way to create an isolate pinned to the audio thread, that would also work, but we don't have any plans to support that.

Alternatively, if you don't need to allow users to run their own Dart code on the audio thread, but just your own library's code, you might want to write that code in C++. We're working on making it easier to ship native code in Dart packages.

Another option would be if miniaudio let you render audio on whatever thread you want (eg by exposing their render loop function, so you can just call it manually). Then you could use ordinary sync callbacks, and render the audio from the target isolate's thread.

**Comment by maks** [comment]
thanks for the suggestions @liamappelbe ! I think I'm probably trying to push things too far generating low latency audio data in Dart, but I'm keen to see how much is possible. 

The temporary Isolates concept sounds very interesting! And having a separate Isolate sounds ok for what I have in mind, but I'm not sure if this will work as I think miniaudio would callback on the some thread that is uses and not the one I call its init function with (ie. a Dart thread invoking FFI) and as you say there is not yet support to pin a specific thread to a Isolate, so I think I need to stick with having to create my own C wrapper lib to be miniaudio's callback that then fetches data to from Dart (eg. some sort of FFI accessed [lockfree ring buffer](https://andrea.lattuada.me/blog/2019/the-design-and-implementation-of-a-lock-free-ring-buffer-with-contiguous-reservations.html)) until that support comes along.

**Comment by dcharkes** [comment]
> using FFI with low-latency audio libs that pretty much all work via callbacks.

On which thread are these callbacks? Is it the same thread that invoked the C code?

And would you need to do anything in Dart with the same data structures besides processing the callbacks?

If it's the same thread, and you have to do nothing else in Dart with the same data structure, you could consider spinning up a new isolate group, and stealing the thread. E.g. `Isolate.spawnUri` -> FFI call to C (never return) -> in C -> many times callback to Dart synchronous with `Pointer.fromFunction`.
Then you never return from that initial call to C.
This would mean that no other Dart code in that Isolate would ever be processed. (Say you had some timer, or some port you were listening to.)
So in some sense it has similar restrictions to the temporary isolate approach, however subsequent callbacks can access the state from previous callbacks.

Tangent: Another idea we had was to compile a subset of Dart that does not interact with the GC and the VM features (say integers, doubles, pointers, FFI calls) to a code style that is independent from Isolates. That would enable arbitrary threads to do callbacks. A prototype is here: https://dart-review.googlesource.com/c/sdk/+/246241 However, after exploring that idea we realized writing that code in C/C++/Rust/.. instead of in Dart is probably more readable and maintainable.

> so I think I need to stick with having to create my own C wrapper lib to be miniaudio's callback that then fetches data to from Dart

Based on the tangent, that might be the right thing to do.

> eg. some sort of FFI accessed [lockfree ring buffer](https://andrea.lattuada.me/blog/2019/the-design-and-implementation-of-a-lock-free-ring-buffer-with-contiguous-reservations.html))

Cool stuff! 😎

**Comment by maks** [comment]
Thanks for all the extra suggestions @dcharkes ! 

Unfortunately I think it is a different thread that the callbacks to fill in the audio buffer use than than init call, the example in their readme titled "...using the low level API." seems to show that's the case (if my C reading is right).

Your tangent however looks **extremely** interesting!! Having that feature available would be brilliant for me as it would also address the need to avoid dynamic memory allocation which is preferably to be avoided for the audio thread (for low latency audio applications). But its not just for this use case that this feature would be really useful but for others that I've looked at as well such as writing small Dart programs to compile into the current non-GC supporting WASM runtimes (which for now is basically *every* wasm runtime bar v8 & spidermonkey behind flags).  

I know previously there has not been interest in supporting subsets of Dart, but I would argue that this would be very valuable. It is a valid question to ask why use Dart for these applications when langs like C & Rust are available and are more suited. My response to this: is that langs like Rust are large and complex and are a significant investment for Dart developers to learn and maintain a working knowledge of when they already have invested in learning Dart for their daily development environment of Flutter. I think at least some Flutter developers who have needed to use Rust for their Flutter projects have gone towards using a package that minimises the hassle of interfacing with Rust and would still need to deal with the cognitive load of dealing with a lang which has not only very different syntax but very different memory management semantics. But again it really doesn't matter if its Rust, Zig or C, it's the 1-thread per Isolate model that means that Dart devs are forced into learning another lang in order to create wrapper libs around the actual native libs they want to use in order to get around the limitation.  

I can see sadly that the patchset was marked as abandoned in favour of #50565 but I think that's unfortunate as while having a better way to bundle native C libs is definitely welcome and needed, it doesn't really address this issue of not being able to do non-Dart thread callbacks into Dart. I appreciate its only a _niche_ use-case, but it's definitely one that people want to address as evidenced by multiple attempts to get it working, for example only yesterday I came across another package (plugin) trying to solve this same use-case, in [this case via some spirited buffer gymnastics in Kotlin](https://github.com/8fdafs2/flutter_raw_sound/blob/main/android/src/main/kotlin/com/codevalop/raw_sound/RawSoundPlayer.kt) which unfortunately I don't think will work so well for low-latency audio given the use of a push vs pull model to filling audio buffers.

Given the code has already been written, is there any chance that the "Isolate Independent Code prototype" could be resurrected, perhaps as an experiment behind a flag? to then allow myself and others to have a chance to try using it to see if the functionality can be used to solve the problem of passing data to callbacks from specific native threads.

**Comment by liamappelbe** [comment]
@maks From the discussions we've had within Dart/Flutter around these features, I think the solution that addresses your use case that's most likely to land is the temporary isolate approach. But I think I didn't do a very good job of explaining it in my previous comment.

From skimming the miniaudio docs, it sounds like the audio is rendered on a dedicated thread created inside the library (referred to as the "audio thread"), which limits your options because you don't have control over the thread. Using the temporary isolate solution, the flow would be:

1. Create a `NativeCallable` from a Dart function using a different constructor that doesn't exist yet (eg `NativeCallable.onTemporaryIsolate` or something), inside an existing isolate (eg the UI isolate).
2. Pass the `NativeCallable.nativeFunction` down to miniaudio.
3. Later, on the audio thread, miniaudio invokes the `nativeFunction`. There is no Dart isolate running on this thread.
4. The Dart VM creates a temporary isolate that runs on the audio thread and runs the target Dart function. (we would probably cache the temporary isolates between invocations, rather than making a new one each time)
5. Miniaudio sees the Dart function run like an ordinary function. It waits for it to finish.

The main drawback of this approach is that the Dart function is running in a separate isolate to the original isolate it was created in (the UI isolate in this example). So accessing/mutating global variables would do unexpected things (in fact we'd probably just disallow accessing global state to be safe). So this is essentially running a subset of Dart, but should be less limiting than the subset Daco mentioned that runs outside of any isolate (it's also a lot easier to implement).

**Comment by maks** [comment]
Thanks for the extra detailed explanation @liamappelbe !

I think I understand the temporary Isolate approach much better now and you're right, it's pretty much what I am asking for. Being in a separate Isolate and not being able to mutate global state shoudn't be an issue for what I have in mind as initially I just want to be able to generate data the for audio samples by experimenting with using[ basic oscillators](https://github.com/maks/dart_synth), [FM](https://github.com/maks/droid-synth), [soundfonts](https://pub.dev/packages/dart_melty_soundfont), etc.

Btw I think you already know this @liamappelbe but its not just miniaudio, I think most OS's native audio subsystems (iOS, MacOS, Linux) have these sort of callback based APIs using a specific "audio render thread" which I think is why miniaudio and other x-platform "wrapper" libs like PortAudio all take this approach.

Of course I'd need some way to get some state data (just readonly, eg. app settings) into the temporary Isolates, perhaps there could be the option to pass in a pointer to the potential new `NativeCallable.onTemporaryIsolate` that would then be available inside the temporary Isolates as well to allow for a very basic mechanism for getting data into the temp Isolate without needing to use `SendPort`s to receive it from "normal" Isolates with the expectation that developer would need to take care of locking around whatever memory that pointer is to.

Having the option to not just cache, but reuse Isolates in this scheme would also be very useful as that way the temp Isolates could maintain local state (eg state of oscillators) for every subsequent callback into the Isolate by the native thread.

I should say that I'm already very happy with the initial `NativeCallable.listener` that you have already created, it already unlocks using a number of other native APIs that I haven't been able to use due to needing to provide native callback functions 👍🏻 , but I'm also excited at the prospect of having an additional temp isolate api that coudl help work with things like audio apis too.

**Comment by mkustermann** [comment]
> Of course I'd need some way to get some state data (just readonly, eg. app settings) into the temporary Isolates, perhaps there could be the option to pass in a pointer to the potential new NativeCallable.onTemporaryIsolate that would then be available inside the temporary Isolates as well to allow for a very basic mechanism for getting data into the temp Isolate without needing to use SendPorts to receive it from "normal" Isolates with the expectation that developer would need to take care of locking around whatever memory that pointer is to.

We already have plans to allow invoking closures in ffi callbacks (see https://github.com/dart-lang/sdk/issues/52689). This would work here as well, as the temporary isolate could invoke a (transitive copy) of the closure. (i.e. `NativeCallable.onTemporaryIsolate(<closure>)` would copy the closure - just as `SendPort.send(<closure>)` would).

> Having the option to not just cache, but reuse Isolates in this scheme would also be very useful as that way the temp Isolates could maintain local state (eg state of oscillators) for every subsequent callback into the Isolate by the native thread.

The caching that we were thinking about is mostly about the data structures the VM has to maintain for an isolate. If the callback is invoked in parallel on multiple threads all threads would get a different temporary isolate.

That being said, it would be conceivable to have a mode where the temporary isolate also maintains state in Dart across invocations - though it would require the VM to guarantee only one native thread can execute the callback pointer at any given point in time.

**Comment by insinfo** [comment]
@dcharkes 
>  Another idea we had was to compile a subset of Dart that does not interact with the GC and the VM features...

That would be very interesting, and I think it has a lot to do with what I requested here. https://github.com/dart-lang/language/issues/1758

I think having a strict mode in dart is somewhat close to C# unsafe which allows running parts of code outside the garbage collector and doing raw pointer operations with a better syntax

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/unsafe

I think it would be very positive to have this strict mode in dart to be able to optimize parts of a code and maybe it would also open up the possibility of having multithreading in this strict mode, that would be fantastic to not need to leave dart to do certain things, without needing the complexity having to rely on another language or cross-platform software development stack with rust or c++

**Comment by Keithcat1** [comment]
How would I keep track of which Dart object my callback was bound to? Can I use a closure?

**Comment by liamappelbe** [comment]
> How would I keep track of which Dart object my callback was bound to? Can I use a closure?

@Keithcat1 At the moment you can only create FFI callbacks from static functions, but we're planning to add closure support soon. The infrastructure we needed to build for async callbacks will also let us support closures: https://github.com/dart-lang/sdk/issues/52689

**Comment by LEggcookies** [comment]
> And would you need to do anything in Dart with the same data structures besides processing the callbacks?

I encountered an obstacle.
![image](https://github.com/dart-lang/sdk/assets/33369299/418d4e36-df9f-4677-84cd-74c699ec0054)


As shown in the diagram above, `device->onIncoming` does not execute within the thread called by `connectDevice`, causing the `receiveCallback` to fail. I need to send all these parameters back to Flutter for processing. I tried using ReceivePorts, but I don't know how to wrap so much content in a Dart_CObject, and how to extract the corresponding parameters from the message in Flutter. I hope to get your help and would be greatly appreciative!

**Comment by dcharkes** [comment]
If you are on a dev release of Dart after 13ec07415b17e8d2da94d0fd7e6df7f216f40d0d, you can use the new `NativeCallable.listener`.

**Comment by nmfisher** [comment]
It seems these aren't usable via `dart2wasm` yet. Any suggestions for callbacks from native code to an `async` Dart function?

**Comment by mraleph** [comment]
>  Any suggestions for callbacks from native code to an `async` Dart function when compiled to wasm?

You seem to be asking the question in the opposite direction. So just to clarify you have a Dart function that returns a `Future` and you want to call it from native code and pretend that this call is synchronous? Can you maybe post an example code of what you are trying to achieve.

**Comment by nmfisher** [comment]
Sorry, that should have been callbacks "to" native code. Exactly what you suggested - calling an async Dart function from native code and either treating the result as synchronous, or at least some other way of propagating a result back to native code that's compatible with wasm. My first attempt was using native Dart ports, but these seem not to be supported under wasm. 

e.g. something like the following (apologies for formatting, I'm on my phone and will clean up when I'm back at a computer).

C (compiled to wasm)
void* (*load_resource_dart)(const char* path); // where this is a function pointer to the Dart function loadResource, assume we've set this somewhere else already 

void load_resource_from_flutter(const char* asset) {
    void* flutterAssetData = load_resource_dart(asset);
}

Dart
Future<Pointer<Void>> loadResource(Pointer<Char> path) async {
 var data = await rootBundle.load(path.toDartString());
// allocate a pointer & memcpy etc
return dataPtr;
}

**Comment by mraleph** [comment]
@nmfisher well, that's not going to work. Consider the following: to call asynchronous function synchronously you need to pause Wasm execution until async function completes then resume it later.

What you can do is to restructure your native code to use callbacks and thread some additional context through them manually:

```
// function pointer to the Dart function loadResource 
// assume we've already set this appropriately   
void* (*_load_resource_dart)(const char* path, void* ctx, void (*loaded)(void* ctx, void* data)); 

// this is a native function that will call an async Dart function
void load_resource_from_flutter(const char* asset) {
    void* ctx = ...; // Some context that needs threading through async code.
    load_resource_dart(asset, ctx, &flutter_resource_loaded);
}

void flutter_resource_loaded(void* ctx, void* flutterAssetData) {
   // Asset loaded.
}
```

**Comment by nmfisher** [comment]
Thanks @mraleph - that's the approach I've been using so far, I was just having issues with concurrency on the native side when using dart2js and then function pointers when using dart2wasm, so I was wondering if there was a Dart-first approach that was better.

I'll keep plugging away and see if I can get it to work as expected.

**Comment by nmfisher** [comment]
Actually I suppose this is a good a place as any to ask - is it even currently possible to use `Pointer.fromFunction` via dart2wasm?

e.g.
```
import "dart:ffi";

void foo() {
  print("foo");
}

void main() {
  var ptr = Pointer.fromFunction<Void Function()>(foo);
  print(ptr);
}
```

`dart --enable-asserts pkg/dart2wasm/bin/dart2wasm.dart foo.dart foo.wasm`
`d8 --experimental-wasm-gc --experimental-wasm-type-reflection pkg/dart2wasm/bin/run_wasm.js -- foo.mjs foo.wasm`

fails with:

```
NoSuchMethodError: method not found: '_nativeCallbackFunction'
Receiver: null
Arguments: []
    at NoSuchMethodError._throwUnimplementedExternalMemberError (wasm://wasm/0008dd2a:wasm-function[188]:0x13763)
    at _nativeCallbackFunction (wasm://wasm/0008dd2a:wasm-function[589]:0x1d0f6)
    at _#ffiCallback0 (wasm://wasm/0008dd2a:wasm-function[585]:0x1d0be)
    at main (wasm://wasm/0008dd2a:wasm-function[56]:0xfb89)
    at main tear-off trampoline (wasm://wasm/0008dd2a:wasm-function[58]:0xfbab)
    at _invokeMain (wasm://wasm/0008dd2a:wasm-function[67]:0x101de)
    at Module.invoke (/tmp/foo.mjs:168:28)
    at main (pkg/dart2wasm/bin/run_wasm.js:352:21)
```

**Comment by dcharkes** [comment]
This issue is currently tracking all kinds of asynchronous use cases.

The basic use case of an asynchronous call is now supported via https://api.dart.dev/stable/3.2.4/dart-ffi/NativeCallable/NativeCallable.listener.html.

Further work for other use cases is tracked in more specific issues:

* Callbacks on the platform thread: https://github.com/flutter/flutter/issues/136314
* Callbacks into already running isolates: https://github.com/dart-lang/language/pull/3531
* Callbacks into temporary isolates https://github.com/dart-lang/sdk/issues/54530

(Side note: all of the above use cases are synchronous from the native code point of view, but asynchronous from the spawning Dart isolate point of view.)

I'll close this issue. Please open a new issue if your use case is not covered by one of the remaining open issues.

Thanks @liamappelbe for implementing `NativeCallable`s! 🎉

**Comment by insinfo** [comment]
Can you use NativeCallable for system audio API callbacks or miniaudio libraries like PortAudio?

**Comment by maks** [comment]
@insinfo you beat me to asking that :-)
though I think the answer is no given that NativeCallable is still async via a Port and maybe this is more aligned with the work being done in[ #136314](https://github.com/flutter/flutter/issues/136314) ?

**Comment by Keithcat1** [comment]
I doubt it. AFAIK NativeCallable won't wait for Dart to run your callback on the main thread, so from Miniaudio's point of view, you it assumes that since the callback has already finished, you aren't using the data anymore and it is free to do something else with it. NativeCallable will give you pointers to Miniaudio buffers, but not whatever data was in those buffers (the buffers can change before you get a chance to read it!). The Miniaudio device event APIs would work, though and Miniaudio Osaudio (miniaudio/extras/osaudio) can be used to record or play audio data without using threads, it'll just block until the ammount you asked for is ready.

**Comment by insinfo** [comment]
It's a little sad and discouraging that we still don't have a pure Dart solution for Audio API callbacks, Dart ffi won't be complete without this, the team behind Dart is very talented and I hope that a solution will be implemented in the future

**Comment by dcharkes** [comment]
> Can you use NativeCallable for system audio API callbacks or miniaudio libraries like PortAudio?

I'm assuming the callbacks (1) come from another thread than the mutator thread of the isolate, and (2) need to be blocking and return an answer. If this is the case one of the following:

> * Callbacks into already running isolates: https://github.com/dart-lang/language/pull/3531
> * Callbacks into temporary isolates https://github.com/dart-lang/sdk/issues/54530

The first one is being discussed. The second one is currently on hold because the first is a more general solution.

**Comment by maks** [comment]
Thanks @dcharkes  for the pointer to the new proposal in the Shared Memory Multithreading PR, very exciting to see it being worked on!

And yes you are correct, most audio APIs seem to follow the model of application code needing to register a callback with the audio system, that then gets called back by a thread (which thread is determined by the audio system) which passes in a pointer to a buffer which it expects to be filled in with required number of audio samples by the time that callback function returns.

So from that proposal PRs doc, this:
>* dealing with APIs which want to call _back_ from native into Dart on an arbitrary thread.

is exactly the case for audio APIs. @dcharkes is commenting on that markdown doc in the PR the best way to contribute to the discussion?

**Comment by dcharkes** [comment]
>  @dcharkes is commenting on that markdown doc in the PR the best way to contribute to the discussion?

Yes.

**Comment by sanjaygholap** [comment]
Hi can you please help me to resolved this issue "cannot invoke a native callback outside of an isolate" Now what happened in the C++ DLL file. Internally, they started a separate thread with a callback. so how to handle this issue.

**Comment by liamappelbe** [comment]
@sanjaygholap That error message means you're using a sync callback (either `NativeCallable.isolateLocal` or `Pointer.fromFunction`). You need to use `NativeCallable.listener` instead.

**Comment by liamappelbe** [comment]
@sanjaygholap It's not possible to hit [that error message](https://github.com/dart-lang/sdk/blob/main/runtime/vm/runtime_entry.cc#L4098) unless you're using a sync callback. Async callbacks follow [this code path](https://github.com/dart-lang/sdk/blob/main/runtime/vm/runtime_entry.cc#L4048) instead, which [always returns](https://github.com/dart-lang/sdk/blob/main/runtime/vm/runtime_entry.cc#L4092) before that error. Async callbacks are specifically designed to work from any thread, so they should work even if your DLL spawned a new thread. Double check that your callback is a `NativeCallable.listener`.

If you're sure that you have a `NativeCallable.listener` that is hitting that error, that would be a new bug. Open a separate issue with repro steps and I'll take a look.

**Comment by sanjaygholap** [comment]
@liamappelbe sir, we get  same issue "Cannot invoke a native callback outside of an isolate".Within the C++ DLL file. A separate thread with a callback was started internally.
I create new issue please check "https://github.com/dart-lang/sdk/issues/55626"
